pkg/tfgen/renames.go

```diff
@@ -0,0 +1,139 @@
+	// type, a datasource (function) type, or an auxiliary type.
+	RenamedProperties map[tokens.Token]map[tokens.Name]string `json:"renamedProperties,omitempty"`
+
+	// Similar to RenamedProperties but pertains to provider-level config.
+	RenamedConfigProperties map[tokens.Name]string `json:"renamedConfigProperties,omitempty"`
+}
+
+func newRenames() Renames {
+	return Renames{
+		Resources:               map[tokens.Type]string{},
+		Functions:               map[tokens.ModuleMember]string{},
+		RenamedProperties:       map[tokens.Token]map[tokens.Name]string{},
+		RenamedConfigProperties: map[tokens.Name]string{},
+	}
+}
+
+func (r Renames) renamedProps(tok tokens.Token) map[tokens.Name]string {
+	props, ok := r.RenamedProperties[tok]
+	if ok {
+		return props
+	}
+	props = map[tokens.Name]string{}
+	r.RenamedProperties[tok] = props
+	return props
+}
+
+type rename struct {
+	parent paths.TypePath
+	name   paths.PropertyName
+}
+
+type renamesBuilder struct {
+	pkg            tokens.Package
+	resourcePrefix string
+	renames        []rename
+	resources      map[tokens.Type]string
+	functions      map[tokens.ModuleMember]string
+	objectTypes    map[string]tokens.Type
+}
+
+func newRenamesBuilder(pkg tokens.Package, resourcePrefix string) *renamesBuilder {
+	return &renamesBuilder{
+		pkg:            pkg,
+		resourcePrefix: resourcePrefix,
+		resources:      map[tokens.Type]string{},
+		functions:      map[tokens.ModuleMember]string{},
+		objectTypes:    map[string]tokens.Type{},
+	}
+}
+
+func (r *renamesBuilder) registerResource(rp *paths.ResourcePath) {
+	if rp.IsProvider() {
+		// Provider resources are not a Terraform concept and can be skipped here.
+		return
+	}
+	r.resources[rp.Token()] = rp.Key()
+}
+
+func (r *renamesBuilder) registerDataSource(dp *paths.DataSourcePath) {
+	r.functions[dp.Token()] = dp.Key()
+}
+
+func (r *renamesBuilder) registerNamedObjectType(paths paths.TypePathSet, token tokens.Type) {
+	if r.objectTypes == nil {
+		r.objectTypes = map[string]tokens.Type{}
+	}
+	for _, p := range paths.Paths() {
+		r.objectTypes[p.String()] = token
+	}
+}
+
+func (r *renamesBuilder) registerProperty(parent paths.TypePath, name paths.PropertyName) {
+	if name.Key == string(name.Name) {
+		return
+	}
+	r.renames = append(r.renames, rename{parent, name})
+}
+
+func (r *renamesBuilder) build() Renames {
+	re := newRenames()
+
+	for _, item := range r.renames {
+		var m map[tokens.Name]string
+		switch parent := item.parent.(type) {
+		case *paths.ConfigPath:
+			m = re.RenamedConfigProperties
+		case *paths.ResourceMemberPath:
+			m = re.renamedProps(tokens.Token(parent.ResourcePath.Token()))
+		case *paths.DataSourceMemberPath:
+			m = re.renamedProps(tokens.Token(parent.DataSourcePath.Token()))
+		default:
+			tok, ok := r.objectTypes[parent.String()]
+			if !ok {
+				panic(fmt.Sprintf("expected registerNamedObjectType to be called for %s", parent))
```

Comment:
https://github.com/pulumi/pulumi-terraform-bridge/blob/a9f77ab0e6c8140e20a53c646303cbb9a0cb6ac2/pkg/tfgen/generate_schema.go#L664-#L664

It's not obvious no, but this is pretty tough to otherwise express. The design is such that object types and token mapping should be done in the first pass. If it's not done by the time we get here something is terribly wrong but this should never happen.